async_hooks: unmerge `resource_symbol` from `owner_symbol`

RaisinTen opened the pull request:
This reverts 7ca2f1303996e0c79c354e979a1527da444ca886 and 937bbc5571073d1fbeffd2d9e18949b3b4dcf09b and adds some regression tests for the referenced issue.

Fixes: https://github.com/nodejs/node/issues/40693

Signed-off-by: Darshan Sen <darshan.sen@postman.com>

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

targos commented:
@nodejs/diagnostics (btw @vdeturckheim would you like to be added to that team?)

RaisinTen commented:
cc @nodejs/async_hooks

vdeturckheim commented:
@targos thanks for the heads up, I am in @nodejs/async_hooks but it probably makes sense for me to join @nodejs/diagnostics too 🤔

targos commented:
@vdeturckheim I don't see you in the async_hooks team either.

orgads commented on test/async-hooks/test-async-local-storage-socket.js:
```diff
@@ -0,0 +1,25 @@
+net
+  .createServer((socket) => {
+    socket.write('Hello, world!');
+    socket.pipe(socket);
```
What does it do?

RaisinTen replied:
@orgads this creates a mock server for the first test case you shared in the issue.

orgads replied:
I meant specifically this line.

```javascript
socket.pipe(socket);
```

Does it echo whatever the socket accepts back to the client?

Qard commented:
Seems I missed #38468...

IMO, `resource_symbol` should never have been removed in the first place. The resource stack logic needs a different object from what `owner_symbol` provides in _many_ cases. Using the constructor name to conditionally trigger additional unwrapping seems a bit fragile to me. What if user code creates a resource with the same name? This can be achieved trivially by sub-classes `AsyncResource`.

Seems to me like it'd be best to revert #38468 and keep those separate, or maybe an owner hierarchy where the _regular_ `owner_symbol` stuff just grabs the direct owner while `executionAsyncResource(...)` keeps following the links until it finds something without an owner?

RaisinTen commented:
@Qard 

> Seems I missed #38468...
> 
> IMO, `resource_symbol` should never have been removed in the first place. The resource stack logic needs a different object from what `owner_symbol` provides in _many_ cases.

Could you please share some use cases that need the `resource_symbol` to be present even though the `owner_symbol` exists?

> Using the constructor name to conditionally trigger additional unwrapping seems a bit fragile to me. What if user code creates a resource with the same name? This can be achieved trivially by sub-classes `AsyncResource`.

In that case, I think https://github.com/nodejs/node/pull/40741/commits/bea58a1fc721f6b57ed4df4025c0bae8cf1872d4 should be fine?

> Seems to me like it'd be best to revert #38468 and keep those separate, or maybe an owner hierarchy where the _regular_ `owner_symbol` stuff just grabs the direct owner while `executionAsyncResource(...)` keeps following the links until it finds something without an owner?

If I'm not wrong, currently the `owner_symbol` does indeed point to the direct owner of a resource. I've submitted https://github.com/nodejs/node/pull/40741/commits/f70bda8ce5e1afe4c180fca026aefe83a7b439bc to let `executionAsyncResource()` return the topmost owner of a resource. With this fix, do we still need to revert https://github.com/nodejs/node/pull/38468?

Flarna commented on lib/async_hooks.js:
```diff
@@ -252,10 +252,65 @@ class AsyncResource {
 const storageHook = createHook({
   init(asyncId, type, triggerAsyncId, resource) {
-    const currentResource = executionAsyncResource();
+    const currentResource = lookupResourceWithStorage();
```
Does this mean the API [`executionAsyncResource()`](https://nodejs.org/dist/latest-v16.x/docs/api/async_hooks.html#async_hooksexecutionasyncresource) are our our of sync now?

RaisinTen replied:
@Flarna in the cases where executionAsyncResource() returns a net/dgram.Socket or a TLSSocket, the AsyncLocalStorage is saved inside the handle, i.e., the TCP/UDP object. Otherwise, it is still in sync.

Flarna replied:
I think it should be always in sync. I don't think that `AsyncLocalStore` should have a different opinion on what is the current resource than other parts of node using AsyncHooks directly (e.g. domains).

RaisinTen replied:
@Flarna I've moved lookupResourceWithStorage() into executionAsyncResource() in this commit: https://github.com/nodejs/node/pull/40741/commits/9d8b2348af329bd31d40ddafc44c5cb5ca179e63, PTAL

addaleax commented:
> > IMO, `resource_symbol` should never have been removed in the first place. The resource stack logic needs a different object from what `owner_symbol` provides in _many_ cases.
> 
> Could you please share some use cases that need the `resource_symbol` to be present even though the `owner_symbol` exists?

I would also be curious about this – both `resource_symbol` and `owner_symbol` point to the public JS API object associated with the current resource that should be returned for diagnostic purposes. They have the same semantics.

Qard commented:
Any sort of connection pooling needs an AsyncResource rather than the public API object to express the individual task, because sharing a resource across multiple tasks will corrupt the context. This happens in http.Agent and many places in userland code. This is why the resource was separate from owner in the first place. It _could_ be _nested_ on the owner though, but I feel like we would want something a bit less fragile than a bunch of constructor name checks which, as I said, can be broken by simply sub-classing AsyncResource, which a bunch of userland already does.

I do agree with the idea of simplifying this as much as we can, but what we're seeing here is that we simplified too much and broke the handle reuse code. We still need to handle that properly somehow, so I feel the removal of resource_symbol needs a bit more thought.

I won't block on this PR, because this needs fixing quickly, but I would request at least a comment with a big, bold WARNING about the safety of constructor name checking and a TODO to improve the handling of that.

RaisinTen commented:
@Qard
> Any sort of connection pooling needs an AsyncResource rather than the public API object to express the individual task, because sharing a resource across multiple tasks will corrupt the context. This happens in http.Agent and many places in userland code. This is why the resource was separate from owner in the first place. It _could_ be _nested_ on the owner though, but I feel like we would want something a bit less fragile than a bunch of constructor name checks which, as I said, can be broken by simply sub-classing AsyncResource, which a bunch of userland already does.

If you're talking about handle reuse, I've added the tests back in https://github.com/nodejs/node/pull/40741/commits/323f8a7e7f035a3838894dcc80baab368357fc7c, which I removed by mistake instead of fixing them in https://github.com/nodejs/node/pull/38468, so I think it still works as expected.
 
> I won't block on this PR, because this needs fixing quickly, but I would request at least a comment with a big, bold WARNING about the safety of constructor name checking and a TODO to improve the handling of that.

Wasn't this already addressed in https://github.com/nodejs/node/commit/bea58a1fc721f6b57ed4df4025c0bae8cf1872d4 or am I missing something here?

Qard commented:
Yep, I had missed those changes when I wrote that comment as I had left the tab open for awhile to review before commenting and it had not updated itself. The current way is _better_ but I'm not a fan of putting a bunch of lazy loading in such a hot function. The `executionAsyncResource` function will be called very frequently so it's best to pull as much of that cost as possible out of band from that call. That's a big part of why the `resource_symbol` property used to be computed ahead of time and only accessed later.

I'm still feeling like reverting the previous change is the better solution. It kind of sucks having two similar purposed properties, but much better for performance than trying to compute that every time it's accessed, which will happen more or at least as much and requires more logic to map back to what it was than just the original storing the resource in a symbol property.

RaisinTen commented:
@Qard

> Yep, I had missed those changes when I wrote that comment as I had left the tab open for awhile to review before commenting and it had not updated itself. The current way is _better_ but I'm not a fan of putting a bunch of lazy loading in such a hot function. The `executionAsyncResource` function will be called very frequently so it's best to pull as much of that cost as possible out of band from that call. That's a big part of why the `resource_symbol` property used to be computed ahead of time and only accessed later.
>
> I'm still feeling like reverting the previous change is the better solution. It kind of sucks having two similar purposed properties, but much better for performance than trying to compute that every time it's accessed, which will happen more or at least as much and requires more logic to map back to what it was than just the original storing the resource in a symbol property.

I ran some benchmarks locally and indeed there was a 23% slow down for `benchmark/async_hooks/async-local-storage-run.js` because of the lazy loads, so I unmerged the symbols, PTAL

Qard reviewed (approved):
Thanks! I'm happy with that. I do encourage you to see if you can come up with a better way though. I'm happy to chat on Slack or Twitter about ideas on how to improve it without being a big performance hit. I'm sure if you don't come up with something, I'll make some changes myself eventually, when I find the time.

Sorry for making you run around in circles a bit on that. It's not normally a big deal doing some per-use checking like that, but this happens to be a function that gets called very, very frequently so it's quite sensitive to performance regressions.

RaisinTen commented:
> Thanks! I'm happy with that. I do encourage you to see if you can come up with a better way though. I'm happy to chat on Slack or Twitter about ideas on how to improve it without being a big performance hit. I'm sure if you don't come up with something, I'll make some changes myself eventually, when I find the time.
>
> Sorry for making you run around in circles a bit on that. It's not normally a big deal doing some per-use checking like that, but this happens to be a function that gets called very, very frequently so it's quite sensitive to performance regressions.

Awesome! I too would like to get some input on ways to merge these symbols properly and I'll also add back the TODO comment that addaleax left because merging the symbols still seems to be the right way forward. Btw, I think you reviewed the outdated commits that I removed during the recent rebase, so I'm resolving those comments. Could you please take a look at the latest changes?

Flarna commented on lib/internal/async_hooks.js:
```diff
@@ -178,13 +178,11 @@ function fatalError(e) {
+  // TODO(addaleax): Merge this with owner_symbol and use it across all
+  // AsyncWrap instances.
+  const publicResource = resource[resource_symbol];
+  if (publicResource !== undefined)
```
```suggestion
  if (publicResource != null)
```

RaisinTen replied:
@Flarna why though? Could `publicResource` be `null` in any case that I'm overlooking?

Flarna replied:
Looked again. Seems I got confused during the review. It's fine as it is.

Qard reviewed (approved):
Still LGTM. 🙂

I'm in Paris this week for work, but I can maybe find some time next week to take a closer look at the symbol merging stuff to see if there's a good way to do it safely. I think I have an idea, but I'll need to look closer--basically have executionAsyncResource get the immediate owner_symbol field and have other places that want the true owner to the while loop thing to unwrap until there's no other specified owner, the handle reuse objects would just have its own owner field mapping to the real owner and not the reuse resource. I _think_ that should work, but I don't have the time to look too closely at it right now. 🤔

orgads commented on test/async-hooks/test-async-local-storage-tlssocket.js:
```diff
@@ -0,0 +1,36 @@
+  .listen(0, function() {
+    const asyncLocalStorage = new AsyncLocalStorage();
+    const store = { val: 'abcd' };
+    asyncLocalStorage.run({ val: 'abcd' }, () => {
```
Replace `{ val: 'abcd' }` with `store`?

orgads commented:
Thank you very much for fixing this.

Is it going to be merged and released soon?

RaisinTen commented:
Landed in 1b8bfddf8908c40c0d94b52e063946589fe548bc, ff3989572acd50419e5b833776b1d2d0546fa616 and 2b0087f3d4df741d8031724c9abd2ea71dcc4dab

orgads commented:
Thank you very much. Will it be backported to v16 automatically?

RaisinTen commented:
@orgads it would normally be done by PRs like https://github.com/nodejs/node/pull/40504. If it doesn't land cleanly, someone will add a `backport-requested` label to this PR and then someone would have to do it manually to resolve the git conflicts.

orgads commented:
Thanks.

willianpc commented:
Hello folks, is there a particular reason why this change was not back ported into v16.x yet?
Are there near future plans to do it? I see that v17 is updated, but nothing landed into v16.x, back ported or not.

Cheers,
- Willian

orgads commented:
When is the next 16 release planned? It's been a month since the last minor release. Previous releases were more frequent.

I couldn't find a release roadmap/schedule anywhere.

richardlau commented:
Currently planned for sometime in January https://github.com/nodejs/Release/issues/658
